api/src/main/java/org/openmrs/Cohort.java

```diff
@@ -141,44 +146,91 @@ public Cohort(String commaSeparatedIds) {
 	 */
 	public String getCommaSeparatedPatientIds() {
 		StringBuilder sb = new StringBuilder();
-		for (Iterator<Integer> i = getMemberIds().iterator(); i.hasNext();) {
-			sb.append(i.next());
-			if (i.hasNext()) {
-				sb.append(",");
-			}
+        for (CohortMembership member : getMembers()) {
+			sb.append(member.getPatient().getPatientId());
+			sb.append(",");
 		}
 		return sb.toString();
 	}
-	
-	public boolean contains(Patient patient) {
-		return getMemberIds() != null && getMemberIds().contains(patient.getPatientId());
-	}
-	
+
+	// TODO: Compare patient objects instead of IDs after removal of deprecated methods
 	public boolean contains(Integer patientId) {
-		return getMemberIds() != null && getMemberIds().contains(patientId);
+		return getMembers() != null && getMembers().stream()
+				.anyMatch(m -> m.getPatient().getPatientId().equals(patientId) &&
+				m.getStartDate().before(new Date()) &&
+				m.getEndDate() == null);
 	}
 	
+	public boolean contains(Patient patient) {
+		return contains(patient.getPatientId());
+	}
+
 	public String toString() {
 		StringBuilder sb = new StringBuilder("Cohort id=" + getCohortId());
 		if (getName() != null) {
-			sb.append(" name=" + getName());
+			sb.append(" name=").append(getName());
 		}
-		if (getMemberIds() != null) {
-			sb.append(" size=" + getMemberIds().size());
+		if (getMembers() != null) {
+			sb.append(" size=").append(getMembers().size());
 		}
 		return sb.toString();
 	}
-	
+
 	public void addMember(Integer memberId) {
-		getMemberIds().add(memberId);
+		this.addMembership(new CohortMembership(new Patient(memberId)));
 	}
-	
-	public void removeMember(Integer memberId) {
-		getMemberIds().remove(memberId);
+
+	public void removeMember(Patient patient, String voidReason) {
+		List<CohortMembership> memberToRemoveList = getMembers().stream()
+				.filter(m -> m.getPatient().getPatientId().equals(patient.getPatientId())).collect(Collectors.toList());
+		for (CohortMembership member : memberToRemoveList) {
+			if (patient.getVoided()) {
+				member.setEndDate(patient.getDateVoided());
+				member.setVoided(patient.getVoided());
+				member.setDateVoided(patient.getDateVoided());
+				member.setVoidedBy(patient.getVoidedBy());
+				member.setVoidReason(voidReason);
+			} else {
+				Date d = new Date();
+				member.setEndDate(d);
+				member.setVoided(true);
+				member.setDateVoided(d);
+				member.setVoidedBy(Context.getAuthenticatedUser());
+				member.setVoidReason(voidReason);
+			}
+		}
 	}
-	
+
+	public void addMembership(CohortMembership cohortMembership) {
+		if (!this.contains(cohortMembership.getPatient().getPatientId())) {
+			getMembers().add(cohortMembership);
```

Comment:
If setCohort has private access in CohortMembership, won't the method be inaccessible from the Cohort class? And I'd removed the setCohort() method since I had added the access="field" attribute to the mapping according your feedback.